### ISSUE TITLE
Touch controller error warning on PineTime

### ISSUE THREAD

**Issue by Arsen6331** [InfiniTimeOrg/InfiniTime#763]
### 

- [X] I searched for similar bug reports and found none was relevant.

### What happened?

Touch controller error warning appears on PineTime on boot

### What should happen instead?

Touch controller error warning should not appear on PineTime

### Reproduction steps

- Install the DFU artifact from #754
- Reboot PineTime

### More details?

This did not happen before installing the artifact from the mentioned PR.
After tapping Proceed, touch works as expected.

### Version

add-motion-service

**Comment by Arsen6331** [comment]
> If you have an issue with a PR, it should be mentioned in the PR instead, but this may not be related to that PR specifically as it doesn't change the code relevant for detecting the error.

Yes, I realized that and believe it has to do with one of the changes for 1.7.0 since the PR was in that milestone. I mentioned the PR because I wanted to make sure there wasn't something wrong there specifically, though that is unlikely.

> I couldn't reproduce the issue. Did you validate the add-motion-service firmware after you flashed it? If not, what was the version that it rolled back to? Did you reset by pressing reset in the validation dialog or by long pressing the button? What is the value of Touch. in the about screen? The error is shown if it is something other than b4.0.1.

Yes, I validated it. All I see in the touch field is `0.0.0`.

**Comment by Arsen6331** [comment]
I got it a couple weeks ago. The touch does work fine (other than the random touch inputs which others also seem to experience).

**Comment by Riksu9000** [comment]
I just got this error and the value is b4.1.1, so the "vendor id" changed.

**Comment by JF002** [comment]
I've just done some tests : I check the ids (chip, vendor firmware) each time the function `Cst816S::GetTouchInfo()` is called. Here are my findings:
 - the IDs are randomly wrong
 - the log `I2C device frozen, reinitializing it!` is also printed randomly
 - The situation **worsens** when the speed of the bus is reduced (from 400 to 100khz)
 - `Cst816S::GetTouchInfo()`  is called **a lot** when the finger touches the display
   - but adding a delay of 50ms after each READ does not fix the issue
 - I reproduce these corrupted reads more often in the game Twos
 - I tried to disable other I²C devices (HR and motion), but that changed nothing to the results of all these tests

So... should we just assume the I²C communication is not reliable? That doesn't look right to me... I know that this is a serial bus and that interference may occur, but in this case, it doesn't really look like random interference (it happens way too often).

Basically, I think the check we implemented in the touch driver to check that a compatible chip is installed is good, but that the communication with the chip is not as reliable as we would like...

**Comment by Avamander** [comment]
Is the hardware configured properly, all the pullups where they should be?

**Comment by geekbozu** [comment]
Device has hardware pull ups on it. 2.7K IIRC which should be more then strong enough for the 400Khz we are operating the bus at.

**Comment by JF002** [comment]
I noticed something interesting : I patched SystemTask to read the registers of the touch controller more often (every 100ms)
 - When I do not touch the display, the controller is in sleep mode and returns errors, that's expected
 - When I touch the display, the controller is awake and returns *mostly* good values, except some bad ones.
 - I notice those bad ones when I touch the panel again before it has the time to go to sleep mode as if it couldn't handle I²C communication and touch at the same time.

IIRC, when I first implemented the touch driver, the controller would trig a single IRQ when the user touched the display. In this IRQ, I would read the register to figure the gesture (tap, swipe,...) out and fetch X/Y values.

Now (I think since the touch handler was revisited and improved by @Riksu9000), a lot of IRQ are triggered when the user touches the display, meaning we read the content of the register while the user touch/release/move their finger.

Is there anything that would prevent the touch controller from communicating correctly while it's touched by the user?

**Comment by kieranc** [comment]
Can you test it with the watch attached to your wrist or otherwise grounded somehow and see if that impacts it? I'm pretty sure this issue is the one I was experiencing with my dev watch and it seemed like that fact it wasn't on a wrist exacerbated the problem.

**Comment by kieranc** [comment]
> Is the hardware configured properly, all the pullups where they should be?

I can't see discrete pullups on the schematic, are they internal to the nrf? Activated?

**Comment by geekbozu** [comment]
So to add some to this, I noticed while reviewing code on #691 that the indicator would jump to MAX when taping the screen....I need to verify but I think the touch screen is also clobbering reads to the accelerometer temporarily.

**Comment by JF002** [comment]
I did an additional test : built a test program that only polls the touch sensor every 50ms. No OS, not IRQ, no mutex, no other devices : 
![image](https://user-images.githubusercontent.com/2261652/140607650-5de42e63-28fe-432f-bcbd-e6533fabe623.png)

Even with this simple code, I still get bad readings on the device IDs when I keep the finger pressed on the display, or when I tap it.

Does this mean that we have to accept that some readings will be erroneous? We can still sort of workaround this, but we will never be able to achieve 100% reliability:
 - Read the chip ID multiple times to ensure that we are reading valid values
 - Check the validity of the touch info (gesture, x/y values) before using them in the software. Currently, the driver CST816S does not do any validity check (https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/src/drivers/Cst816s.cpp#L70). In case if invalid data, we could end up with X/Y values > 240 (out of screen) and with a invalid gesture.

**Comment by JF002** [comment]
I implemented those checks in https://github.com/InfiniTimeOrg/InfiniTime/pull/811.

I'm not sure if it's related, but I noticed I could easily crash (watchdog reset) InfiniTime when playing *Twos*. The execution was stuck in `TwiMaster::Read()` or `TwiMaster::Write()`, waiting for the STOP event that never came. 

Note that the callstack showed that these blocking call where always coming from `SystemTask::UpdateMoin()`. Removing this call allowed me to play the game for several minutes without a crash.
This is not new (https://github.com/InfiniTimeOrg/InfiniTime/issues/79), but we have never found a solution that was completely reliable...

**Comment by Riksu9000** [comment]
> Check the validity of the touch info (gesture, x/y values) before using them in the software

Did you encounter bad touchinfo values?

If the IDs are wrong when touching, couldn't we check that the touching bit is zero?

**Comment by JF002** [comment]
I added a log to show when bad values were detected but didn't log the values so I cannot tell how corrupted they were, unfortunately. That would be an interesting test!
Also, this is very confusing, but I noticed that I would experience far more issues on 2 devkits than on my sealed unit. For example, the devkits crashes after a few moves in Twos (and I could see a lot of touch/I²C errors in the log), while I could play several minutes on my sealed pinetime without issues (but no log on SWD, as it's sealed).

**Comment by kieranc** [comment]
> Also, this is very confusing, but I noticed that I would experience far more issues on 2 devkits than on my sealed unit. For example, the devkits crashes after a few moves in Twos (and I could see a lot of touch/I²C errors in the log), while I could play several minutes on my sealed pinetime without issues (but no log on SWD, as it's sealed).

In this context, is the sealed device on your wrist and the devkits on the desk/otherwise mounted?

**Comment by StarGate01** [comment]
I noticed that the CST816S in my watch (sometimes) ignores the reset signal. Instead I had to defer the register initialization to the first invocation of the touch event interrupt callback, because at this point the I2C communication window is open.

**Comment by JF002** [comment]
@StarGate01 Do you mean that the reset signal sent on the reset pin (https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/src/drivers/Cst816s.cpp#L23) does not actually resets the chip?

**Comment by StarGate01** [comment]
@JF002 Yes, that is exactly what I suspect. I noticed this behavior because on my pinetime dev kit, reading the chip ID on firmware boot failed >90% of the time. It only worked reliably when I furiously tapped the screen during startup. The datasheet states that the Cst816S has an activity window (default 2 seconds) after a touch event, and I2C is only active during that window. The datasheet recommends performing the exact reset procedure as was implemented already to ensure the window is active, however this did not work reliably for some reason on my device. It might be something about a potentially cloned touch chip, some weird electrostatic effects, capacitance on the signal lines or straight up a hardware defect on my device. However, some users reported the touch driver failing to initialize very sporadically in the past as well.

I notices that upon a failed reset signal not only the chip ID could not be read, but also the interrupt register failes to be written to. This resulted in an incorrect event reporting behavior on my device. My theory of the reset failing is further supported by the fact that the modified configuration state of the touch chip survives firmware reboots, but resets to its default value after a power loss.

Interestingly, the old reset procedure works great on my P8b watch, which uses a Cst716 touch chip. However, that chip does not feature auto-sleep and auto-wake, but instead has an always-on I2C bus (until you manually put it to sleep). Why the actual chip on the PCB is marked Cst816 is another mystery, it reports and behaves exactly like a Cst716.

**Comment by StarGate01** [comment]
@JF002 Yes, that is my workaround, see https://github.com/StarGate01/InfiniTime/commit/402ebaebedfee0e5846842fc8743959eae855a31#diff-9d9d5389cf4cc0c3064d0778c604a6c4f06329d64e4c18a54464a7cbba9682d2= .

I suspect there could be some residual capacitance pulling the reset pin high-ish, as some users reported the touch panel working again after they grounded the aluminium case by touching it.
